### ISSUE TITLE
`firebase deploy` without firebase.json

### ISSUE THREAD

**Issue by Nowaker** [firebase/firebase-tools#213]
We have a project that generated four different artifact-websites, each of these has to be deploy to Firebase. We could generate a separate firebase.json file for each but I don't think it makes much sense. firebase.json is used to generate an archive that's pushed to Firebase. Why not let us provide everything that's configurable in JSON file via parameters to `firebase deploy`?

Here's an example - this `firebase deploy` call is equivalent to having a JSON file.

```bash
firebase deploy \
  --firebase project-name \
  --public ./build
  --ignore firebase.json \
  --ignore '**/.*'
  --ignore '**/node_modules/**'
```

```
{
  "firebase": "project-name",
  "public": "./build",
  "ignore": [
    "firebase.json",
    "**/.*",
    "**/node_modules/**"
  ]
}
```

**Comment by Nowaker** [comment]
Well, I discovered the minimum `firebase.json` file that `firebase deploy` will require is:

```
% cat firebase.json
{}
```
And the deploy command is `firebase deploy --project projectname --public . --only hosting`.

**Comment by mbleigh** [comment]
I think there's more we can do here, but I don't think exposing dozens of command-line parameters is necessarily the way. I'll leave this open while we think.

**Comment by Nowaker** [comment]
At the very minimum, don't force the existence of `firebase.json` if `--project` and `--public` are provided.

**Comment by vtellier** [comment]
+1

I felt on this issue while looking for a way to have a slightly different configuration between my 'deploy' (built source code that I'll push in prod) and 'serve' (non built source code that I want to test quickly in my dev environment):
- **firebase deploy**: take the firebase.json and thats' it
- **firebase serve**: take a --public argument that is not the final built folder, but the dev workspace ; so I can code without having to rebuild my whole project every time I want to refresh my web app.

There is probably a way to achieve this, this kinda 'firebase.json overriting' looks interesting to me.

**Comment by eirikb** [comment]
I think this is a great idea.
@Nowaker in your initial example you use `--ignore '**/node_modules/**'`, but none of the ignores there would be necessary if _build_ is your public folder, right?

**Comment by Nowaker** [comment]
I translated the JSON syntax into command line parameter syntax. If it's not necessary in one, it's not necessary in the other.

**Comment by mbleigh** [comment]
It's going to be extremely difficult to enumerate the entire array of possible config as command-line options. I'd be more amenable to being able to pass the config JSON inline:

    firebase deploy --config '{"hosting":{"public":"dist"}}'

WDYT?

**Comment by vtellier** [comment]
@mbleigh I agree about the entire array enumeration, however if using a JSON string as configuration, it can be painful if we _have to_ specify the full configuration.
Passing JSON that just overrides only the settings it contains, without discarding the actualy configuration file sounds just fine. Discarding the actual configuration file could be maybe optional?

**Comment by eirikb** [comment]
For my own personal sake I only need `--public`, and nothing else. - I only want to get rid of a file I actually don't need.  

However, if I had to specify a lot of configuration, but for different environments, I would prefer specifying the name of a configuration file, .e.g, `--config firebase-dev.json`.  
Or perhaps `--config dev`, and _dev_ is a property under ` {"firebase": { "dev": { ... } }` in _package.json_.

**Comment by m3hdiii** [comment]
@mbleigh thanks for your answer, it helped me to make my project up and running upon firebase.
Just a simple correction :
`firebase deploy --project projectname --public .`

In the above code, we should have use projectId instead of projectname.

so the final line of code for me was something like this:
`firebase deploy --project projectId --public .`

Thanks again

**Comment by Nowaker** [comment]
@mbleigh While you think about a preferable solution, can you *at least* not require the existence of config file if the above command line parameters are. specified? Thanks.

**Comment by mhuebert** [comment]
I would be happy with an ability to specify an alternate config file. When serving locally, I'd like to redirect requests to an HTML file with different API credentials for third-party services.

**Comment by ashgeo** [comment]
First got to project folder 
>>ng build 
above command will generate dist folder which is public path 

{
  "firebase": "project name",
  "public": "./dist",
  "ignore": [
    "firebase.json",
    "**/.*",
    "**/node_modules/**"
  ]
}

**Comment by Nowaker** [comment]
@ashgeo How does it add anything to this topic?

**Comment by cristiano-belloni** [comment]
Any news on this? Taking the path to an alternative file doesn't seem an incredibly complicated feature and would be much more CI-friendly.

**Comment by michaelkowalski** [comment]
Hello, just to add my own way of dealing with this:

1. I created two files: `firebase.development.json` and `firebase.production.json`
2. I created two scripts for deployment and they start with copying the right config to firebase.json, for example: `cp firebase.development firebase.json`

This way I can have multiple configs even if it's not officially supported
